### PR TITLE
Rewrite to use individual state functions

### PR DESCRIPTION
**Pull request by mvpatel2000**
# What does this PR do?

Rewrite to use individual state functions. Splits out model and optimizer state dict functions

### CODE REVIEW

File: composer/trainer/trainer.py [hunk 1]
```diff
@@ -1443,7 +1443,7 @@ def __init__(
                 # Set load arguments to defaults as this applies only to the initial non-autoresume
                 # load. We do not reset `load_strict_model_weights` for models with frozen layers.
                 load_weights_only = False
-                load_ignore_keys = None
+                # load_ignore_keys = None  # TODO: Re-enable load ignore keys
```
Comment (dakinggg) [inline composer/trainer/trainer.py:1446]: reminder to undo

File: composer/core/state.py [hunk 2]
```diff
@@ -874,58 +874,68 @@ def get_model_state_dict(self) -> Dict[str, Any]:
...
+                raise NotImplementedError(
+                    textwrap.dedent(f'fsdp_state_dict_type={self.fsdp_state_dict_type} is not supported for '
+                                    f'torch version {{version.parse(torch.__version__)}} > 2.1.3. Please set '
+                                    'fsdp_state_dict_type to None, "full", or "sharded".'))
 
-    def _legacy_get_model_state_dict(self) -> Dict[str, Any]:
-        if self.fsdp_enabled and self.fsdp_state_dict_type is not None:
-            with fsdp_state_dict_type_context(self.model, state_dict_type=self.fsdp_state_dict_type):
-                model_state_dict = self.model.state_dict()
+            model_state_dict = get_model_state_dict(
```
Comment (dakinggg) [inline composer/core/state.py:885]: does this function work for non distributed training? If so...awesome

Comment (dakinggg) [reply composer/core/state.py:885]: and for ddp does it consume the `module` prefix as we do below?

File: composer/core/state.py [hunk 3]
```diff
@@ -874,58 +874,68 @@ def get_model_state_dict(self) -> Dict[str, Any]:
...
-        if self.fsdp_enabled and self.fsdp_state_dict_type is not None:
-            optim_state_dict = {
-                type(optimizer).__qualname__:
-                    fsdp_get_optim_state_dict(self.model, optimizer, state_dict_type=self.fsdp_state_dict_type)
-            }
-        else:
-            optim_state_dict = {type(optimizer).__qualname__: optimizer.state_dict()}
-        return optim_state_dict
+    def get_optim_state_dict(self) -> Dict[str, Any]:
+        """Collect the state dict for the model.
```
Comment (dakinggg) [inline composer/core/state.py:905]:
```suggestion
        """Collect the state dict for the optimizer.
```

File: composer/core/state.py [hunk 4]
```diff
@@ -874,58 +874,68 @@ def get_model_state_dict(self) -> Dict[str, Any]:
...
 
-    def get_model_and_optimizer_state_dict(self, model_only=False) -> Tuple[Dict[str, Any], Dict[str, Any]]:
+        Returns:
+            Dict[str, Any]: The state dict for the model.
```
Comment (dakinggg) [inline composer/core/state.py:908]:
```suggestion
            Dict[str, Any]: The state dict for the optimizer.
```

File: composer/core/state.py (hunk 2 above)
Comment (mvpatel2000) [reply composer/core/state.py:885]: Yes to non distributed, they are unifying interface.

No to consuming. Good catch will add

File: composer/trainer/trainer.py (hunk 1 above)
Comment (mvpatel2000) [reply composer/trainer/trainer.py:1446]: Leaving for @abhi-mosaic. I have a reminder to revert pre release

**Review by abhi-mosaic — APPROVED** [review]
Approving to unblock. I see the commented out `load_ignore_keys` line, does this mean I can't use `load_ignore_keys` anymore with this branch?

**Pull request merged**